### ISSUE TITLE
ansible: environment difference while running debops playbooks

### ISSUE THREAD

**Issue by dw** [dw/mitogen#122]
My bashrc environment sets en_GB.UTF-8. When running with Mitogen, the run completes, but with non-Mitogen, then run fails with:

````
TASK [debops.sshd : Check if OpenSSH is installed (or half-installed)] *************************************************************************************************************
fatal: [35.200.184.46]: FAILED! => {"changed": false, "failed": true, "failed_when_result": true, "rc": 0, "stderr": "Shared connection to 35.200.184.46 closed.\r\n", "stdout": "bash: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)\r\ninstalled\r\n", "stdout_lines": ["bash: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)", "installed"]}
````

Because the Ansible master's environment is getting copied across somehow, and the locale data was missing on the target box. Figure out what's happening

**Comment by stefangweichinger** [comment]
had that debops.sshd issue as well here. Just fyi.

**Comment by drybjed** [comment]
If it works with Mitogen (I wonder how) but doesn't work without it, it seems like the `dpkg` command doesn't like to use non-existent locales. Your `en_gb.UTF-8` locale is passed through SSH because by default `LC_*` environment variables are whitelisted by `sshd`.

Perhaps setting the environment as `LC_ALL: C` for this task could solve it.

**Comment by dw** [comment]
I just pushed issue_122.yml which I was using earlier to try and reproduce the difference. Running vimdiff on the output with/without Mitogen shows no environment difference. Probably I'm still doing something wrong to reproduce the original scenario, will look at it later in the week.

No matter what the current behaviour is, Mitogen needs to emulate it. It needs to let both the SSH invocation and the sudo invocation do whatever horridness is done to the environment and reproduce exactly that, otherwise more playbooks will be broken behind closed doors

**Comment by drybjed** [comment]
@dw Ah, so you want to make sure that Mitogen fails the same way that Ansible fails with the same environment, good idea.

**Comment by dw** [comment]
Discoveries so far:

- OS X ssh_config does not have SendEnv by default
- Debian ssh_config does not have SendEnv by default
- Fedora ssh_config does not have SendEnv by default, yet somehow the variables are being sent (patch?)

Man, Ansible's integration with everything is so bloody haphazard.. as usual the correct behaviour isn't exactly clear. I dislike having bits of the environment leak into the tooling (it makes it _impossible_ to test), but we have to emulate Ansible *sigh*

**Comment by dw** [comment]
@stefangweichinger do you still have access to your setup that triggered the locale error? If so, can you please --start-at="Check if OpenSSH is installed (or half-installed)" and attach "-vvv" output both for Ansible and for Mitogen.

If not don't worry, I'll pick this up next week

**Comment by stefangweichinger** [comment]
@dw yes, I have. pls more details: with or without your new debug variable exported, mitogen.log from controller?

**Comment by dw** [comment]
Just -vvv is fine, MITOGEN_ROUTER_DEBUG is only useful for lower level IO issues

**Comment by stefangweichinger** [comment]
ansible-playbook ~/.local/share/debops/debops/ansible/playbooks/common.yml  -vvv --start-at="Check if OpenSSH is installed (or half-installed)"  -l  tx100 > debops_common.log

[debops_common.log](https://github.com/dw/mitogen/files/1800742/debops_common.log)

can't find a mitogen.log in /tmp now, though
